pkg/rke2/np.go

```diff
@@ -0,0 +1,113 @@
+package rke2
+
+import (
+	"context"
+	"fmt"
+
+	daemonsConfig "github.com/rancher/k3s/pkg/daemons/config"
+	"github.com/rancher/spur/cli"
+	"github.com/sirupsen/logrus"
+	v1 "k8s.io/api/networking/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/util/retry"
+)
+
+const (
+	namespaceAnnotationNetworkPolicy = "np.rke2.io"
+
+	defaultNetworkPolicyName = "default-network-policy"
+)
+
+// networkPolicy specifies a base level network policy applied
+// to the 3 primary namespace: kube-system, default, and kube-public.
+// This policy only allows for all inter-namespace traffic.
+var networkPolicy = v1.NetworkPolicy{
+	ObjectMeta: metav1.ObjectMeta{
+		Name: defaultNetworkPolicyName,
+	},
+	Spec: v1.NetworkPolicySpec{
+		PodSelector: metav1.LabelSelector{}, // empty to match all pods
+		PolicyTypes: []v1.PolicyType{v1.PolicyTypeIngress},
+		Ingress: []v1.NetworkPolicyIngressRule{
+			{
+				From: []v1.NetworkPolicyPeer{
+					{
+						PodSelector: &metav1.LabelSelector{}, // empty to match all pods
+					},
+				},
+			},
+		},
+	},
+}
+
+// setNetworkPolicy applies the default network policy for the given namespace and updates
+// the given namespaces' annotation.
+func setNetworkPolicy(ctx context.Context, namespace string, cs *kubernetes.Clientset) error {
+	ns, err := cs.CoreV1().Namespaces().Get(ctx, namespace, metav1.GetOptions{})
+	if err != nil {
+		return fmt.Errorf("networkPolicy: get %s - %w", namespace, err)
+	}
+	if ns.Annotations == nil {
+		ns.Annotations = make(map[string]string)
+	}
+	if _, ok := ns.Annotations[namespaceAnnotationNetworkPolicy]; !ok {
+		_, err := cs.NetworkingV1().NetworkPolicies(namespace).Get(ctx, defaultNetworkPolicyName, metav1.GetOptions{})
+		if err != nil {
+			if apierrors.IsAlreadyExists(err) {
+				if err := cs.NetworkingV1().NetworkPolicies(namespace).Delete(ctx, defaultNetworkPolicyName, metav1.DeleteOptions{}); err != nil {
+					return err
+				}
+			} else if !apierrors.IsNotFound(err) {
+				return err
+			}
+		}
+		if _, err := cs.NetworkingV1().NetworkPolicies(namespace).Create(ctx, &networkPolicy, metav1.CreateOptions{}); err != nil {
```

Comment:
So the semantics here are intentional replacement instead of "patching" via UPDATE?

Reply:
To be clear, this is fine if that is what is required. A comment to that effect would be :+1:

Reply:
agree on the comment. answering intention here: yes, intention is that if the annotation is not present, this is either a fresh cluster or the user is explicitly asking for a "reset" of net policy. So, just nuke and pave.